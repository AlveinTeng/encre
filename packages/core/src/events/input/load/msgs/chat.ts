import { BaseMessage, BaseMessageFields, MessageRole } from './base.js';

export interface ChatMessageFields extends BaseMessageFields {
  role: string;
}

export class ChatMessage extends BaseMessage {
  _isSerializable = true;

  static _name(): string {
    return 'ChatMessage';
  }

  role: string;

  constructor(content: string, role: string);

  constructor(fields: ChatMessageFields);

  constructor(fields: string | ChatMessageFields, role?: string) {
    if (typeof fields === 'string') {
      fields = { content: fields, role: role! as MessageRole };
    }
    super(fields);

    this.role = fields.role;
  }

  _role(): MessageRole {
    return 'general';
  }

  concat(message: ChatMessage): ChatMessage {
    return new ChatMessage({
      content: ChatMessage._mergeContent(this.content, message.content),
      additionalKwargs: ChatMessage._mergeAdditionalKwargs(
        this.additionalKwargs as NonNullable<
          BaseMessageFields['additionalKwargs']
        >,
        message.additionalKwargs as NonNullable<
          BaseMessageFields['additionalKwargs']
        >
      ),
      role: this.role,
    });
  }

  static isInstance(message: BaseMessage): message is ChatMessage {
    return message._role() === 'general';
  }
}
