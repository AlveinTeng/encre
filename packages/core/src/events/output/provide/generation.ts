export interface Generation {
  /**
   * Generated output
   */
  output: unknown;

  /**
   * Raw generation information from the provider.
   * May include reason for finishing.
   */
  info?: Record<string, unknown>;
}

export type GenerationChunkField = Generation;

export abstract class GenerationChunk implements Generation {
  public output: unknown;

  public info?: Record<string, unknown>;

  constructor(fields: GenerationChunkField) {
    this.output = fields.output;
    this.info = fields.info;
  }

  abstract concat(chunk: GenerationChunk): GenerationChunk;
}