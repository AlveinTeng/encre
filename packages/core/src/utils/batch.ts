export function batch<T>(arr: T[], size: number): T[][] {
  return arr.reduce((chunks: T[][], element: T, idx: number) => {
    const chunkIdx: number = Math.floor(idx / size);
    const chunk: T[] = chunks[chunkIdx] || [];
    chunks[chunkIdx] = chunk.concat([element]);
    return chunks;
  }, [] as T[][]);
}
