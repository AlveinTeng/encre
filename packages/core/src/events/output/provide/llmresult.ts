import { Generation } from './generation';

/**
 * Contains all relevant information returned by an LLM.
 */
export type LLMResult = {
  /**
   * List of the things generated. Each input could have multiple {@link Generation}, hence this is a list of lists.
   */
  generations: Generation[][];

  /**
   * LLM-provider specific output.
   */
  llmOutput?: Record<string, any>;
}