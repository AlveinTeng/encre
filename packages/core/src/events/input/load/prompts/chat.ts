import { BaseMessage } from '../msgs/base.js';
import { getChatString } from '../msgs/utils.js';
import { BasePrompt } from './base.js';

export interface ChatPromptFields {
  messages: BaseMessage[];
}

export class ChatPrompt extends BasePrompt {
  _isSerializable = true;

  static _name(): string {
    return 'ChatPrompt';
  }

  messages: BaseMessage[];

  constructor(fields: BaseMessage[] | ChatPromptFields) {
    if (Array.isArray(fields)) {
      fields = { messages: fields };
    }

    super(fields);
    this.messages = fields.messages;
  }

  _promptType(): string {
    return 'chat';
  }

  toString(): string {
    return getChatString(this.messages);
  }

  toChatMessages(): BaseMessage[] {
    return this.messages;
  }
}
