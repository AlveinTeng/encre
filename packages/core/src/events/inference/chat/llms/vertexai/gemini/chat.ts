import {
  GoogleGenerativeAI as GoogleGenerativeAIClient,
  Content,
  ModelParams,
  GenerationConfig,
  SafetySetting,
  GenerateContentCandidate,
  GenerateContentResponse,
  GenerativeModel,
  GenerateContentRequest,
  EnhancedGenerateContentResponse,
  GenerateContentResult,
  CountTokensResponse,
  TextPart,
  Part,
} from '@google/generative-ai';

import {
  SecretFields,
  SerializedFields,
} from '../../../../../../load/keymap.js';
import { AsyncCallError } from '../../../../../../utils/asyncCaller.js';
import { getEnvironmentVariables } from '../../../../../../utils/environment.js';
import { BaseMessage } from '../../../../../input/load/msgs/base.js';
import {
  Generation,
  GenerationChunk,
} from '../../../../../output/provide/generation.js';
import { LLMResult } from '../../../../../output/provide/llmresult.js';
import {
  ChatGeneration,
  ChatGenerationChunk,
} from '../../../../../output/provide/message.js';
import { BaseChatLM, BaseLLMParams } from '../../../base.js';
import { TokenUsage } from '../../../index.js';
import {
  GeminiCallOptions,
  GeminiContent,
  GeminiContentRole,
  GeminiInput,
  GeminiSafetySetting,
  checkModelForGemini,
  checkModelForGeminiVision,
  wrapGoogleGenerativeAIClientError,
} from '../index.js';
import {
  getContentFromMessage,
  getMessageFromContent,
  getMessageFromContentWithRole,
  isModalityRequiredInMessage,
} from './utils.js';

export interface GeminiChatParamsBase extends ModelParams {
  contents: Content[];
}

export interface GeminiChatParamsNonStreaming extends GeminiChatParamsBase {
  stream: false;
}

export interface GeminiChatParamsStreaming extends GeminiChatParamsBase {
  stream: true;
}

export type GeminiChatParams =
  | GeminiChatParamsNonStreaming
  | GeminiChatParamsStreaming;

export class GeminiChat<
    CallOptions extends GeminiCallOptions = GeminiCallOptions,
  >
  extends BaseChatLM<CallOptions>
  implements GeminiInput
{
  _isSerializable = true;

  get _secrets(): SecretFields | undefined {
    return {
      googleApiKey: 'GOOGLE_API_KEY',
    };
  }

  get _aliases(): SerializedFields | undefined {
    return {
      googleApiKey: 'apiKey',
    };
  }

  static _name(): string {
    return 'GeminiChat';
  }

  /**
   * `gemini-pro` or `gemini-pro-vision`
   */
  modelName = 'gemini-pro';

  /**
   * Default for gemini-pro: 0.9
   * Default for gemini-pro-vision: 0.4
   */
  temperature = 0.9;

  /**
   * Default: 1.0
   */
  topP = 1.0;

  /**
   * This value must be 1.
   */
  candidateCount = 1;

  /**
   * Default for gemini-pro: 2048
   * Default for gemini-pro-vision: 4096
   */
  maxOutputTokens = 2048;

  /**
   * Whether the response comes with stream
   */
  streaming = true;

  /**
   * Default for gemini-pro: none
   * Default for gemini-pro-vision: 32
   */
  topK: number;

  googleApiKey?: string;

  stopSequences?: string[];

  additionalKwargs?: Record<string, unknown>;

  contents: GeminiContent[];

  safetySettings?: Array<GeminiSafetySetting>;

  private _client: GoogleGenerativeAIClient;

  private _isMultimodal = false;

  _llmType(): string {
    return 'gemini';
  }

  constructor(fields?: Partial<GeminiInput> & BaseLLMParams) {
    super(fields ?? {});

    this.googleApiKey =
      fields?.googleApiKey ?? getEnvironmentVariables('GOOGLE_API_KEY');

    if (!this.googleApiKey) {
      throw new Error('Google API Key not found');
    }

    this.modelName = fields?.modelName ?? this.modelName;

    if (!checkModelForGemini(this.modelName)) {
      throw new Error(
        'model is not valid for Gemini, please change it to `gemini-pro` or `gemini-pro-vision`'
      );
    }

    if (checkModelForGeminiVision(this.modelName)) {
      this._isMultimodal = true;
      this.temperature = 0.4;
      this.topK = fields?.topK ?? 32;
      this.maxOutputTokens = fields?.maxOutputTokens ?? 4096;

      if (this.maxOutputTokens > 4096) {
        console.warn(
          'gemini-pro-vision does not support output token larger than 4096, now using 4096 as maxOutputTokens.'
        );
        this.maxOutputTokens = 4096;
      }
    } else {
      this.topK = fields?.topK ?? 1;
      this.maxOutputTokens = fields?.maxOutputTokens ?? 2048;

      if (this.maxOutputTokens > 2048) {
        console.warn(
          'gemini-pro does not support output token larger than 2048, now using 2048 as maxOutputTokens.'
        );
        this.maxOutputTokens = 2048;
      }
    }

    this.temperature = fields?.temperature ?? this.temperature;
    this.topP = fields?.topP ?? this.topP;

    this.additionalKwargs = fields?.additionalKwargs ?? {};

    this.stopSequences = fields?.stopSequences;
    this.streaming = fields?.streaming ?? this.streaming;
  }

  protected _getGenerationConfig(): GenerationConfig {
    return {
      temperature: this.temperature,
      topP: this.topP,
      topK: this.topK,
      candidateCount: this.candidateCount,
      maxOutputTokens: this.maxOutputTokens,
      stopSequences: this.stopSequences ? this.stopSequences : [],
    };
  }

  // protected _getTools(
  //   options?: this['SerializedCallOptions'] | undefined
  // ): Array<GeminiTool> {
  //   if (options && options['tools']) {
  //   }

  //   return options && options['tools'] ? options['tools'] : [];
  // }

  protected _getSafetySettings(
    options?: this['SerializedCallOptions'] | undefined
  ): Array<SafetySetting> {
    if (options && options['safetySettings']) {
      return options['safetySettings'].map((setting: GeminiSafetySetting) => ({
        category: setting.category,
        threshold: setting.threshold,
      }));
    }

    return [];
  }

  protected _failedAttemptHandler(e: Error): void {
    const error: Error = wrapGoogleGenerativeAIClientError(e);
    super._failedAttemptHandler(error);
  }

  getParams(
    options?: this['SerializedCallOptions']
  ): Omit<GeminiChatParams, 'contents'> {
    return {
      model: this.modelName,
      // tools: this._getTools(options),
      safetySettings: this._getSafetySettings(options),
      generationConfig: this._getGenerationConfig(),
      stream: this.streaming,
      ...this.additionalKwargs,
    };
  }

  async _provide(
    messages: BaseMessage[],
    options: this['SerializedCallOptions']
  ): Promise<LLMResult> {
    if (
      this._isMultimodal !== true &&
      messages.some((message: BaseMessage) =>
        isModalityRequiredInMessage(message)
      )
    ) {
      throw new Error(
        'Message contains Image input but modality is not enabled for gemini-pro, please change it to gemini-pro-vision.'
      );
    }

    const tokenUsage: TokenUsage = {};

    const params = this.getParams(options);
    const convertedContents: Content[] = messages.map((message: BaseMessage) =>
      getContentFromMessage(message)
    );

    if (this.streaming) {
      const stream = this._completionWithStream(
        params,
        convertedContents,
        options
      );

      const chunks: Record<number, ChatGenerationChunk> = {};
      for await (const chunk of stream) {
        const index: number = (chunk.info?.completion as number) ?? 0;
        if (chunks[index] === undefined) {
          chunks[index] = chunk;
        } else {
          chunks[index] = chunks[index].concat(chunk);
        }
      }

      const generations: ChatGenerationChunk[] = Object.keys(chunks)
        .sort((a: string, b: string) => parseInt(a, 10) - parseInt(b, 10))
        .map((key: string): ChatGenerationChunk => chunks[key]);

      const messageTokenUsage = await this.getNumTokensInChat(messages);

      const generationsTokenUsage =
        await this.getNumTokensInGenerations(generations);

      tokenUsage.promptTokens = messageTokenUsage;
      tokenUsage.completionTokens = generationsTokenUsage;
      tokenUsage.totalTokens = messageTokenUsage + generationsTokenUsage;

      return { generations, llmOutput: { tokenUsage } };
    } else {
      const response = await this.completionWithRetry(
        {
          ...params,
          stream: false,
          contents: convertedContents,
        },
        {
          ...options,
        }
      );

      const promptTokens: number = await this.getNumTokensInChat(messages);
      let completionTokens: number | undefined;

      if (!response.candidates) {
        throw new Error('No candidates from Gemini response.');
      }

      const generations: Array<unknown> = [];
      for (const candidate of response.candidates) {
        const numTokensResponse: CountTokensResponse = await this._getClient()
          .getGenerativeModel({
            model: this.modelName,
          })
          .countTokens({
            contents: [candidate.content],
          });

        completionTokens =
          (completionTokens ?? 0) + numTokensResponse.totalTokens;

        const output: string = candidate.content.parts
          .map((part: Part) => (part as TextPart).text)
          .join('\n');

        const message: BaseMessage = getMessageFromContent(
          candidate.content ?? { role: 'assistant' }
        );

        const info = {
          finishReason: candidate.finishReason,
          completion: candidate.index ?? 0,
          finishMessage: candidate.finishMessage,
          safetyRatings: candidate.safetyRatings,
          citationMetadata: candidate.citationMetadata,
        };

        generations.push({
          output,
          message,
          info,
        });
      }

      const totalTokens: number = promptTokens + (completionTokens ?? 0);

      if (completionTokens) {
        tokenUsage.completionTokens =
          (tokenUsage.completionTokens ?? 0) + completionTokens;
      }

      if (promptTokens) {
        tokenUsage.promptTokens = (tokenUsage.promptTokens ?? 0) + promptTokens;
      }

      if (totalTokens) {
        tokenUsage.totalTokens = (tokenUsage.totalTokens ?? 0) + totalTokens;
      }

      return {
        generations: generations as Generation[],
        llmOutput: { tokenUsage },
      };
    }
  }

  private async *_completionWithStream(
    params: Omit<GeminiChatParams, 'contents'>,
    contents: Content[],
    options: this['SerializedCallOptions']
  ): AsyncGenerator<ChatGenerationChunk> {
    const stream: AsyncIterable<GenerateContentResponse> =
      await this.completionWithRetry(
        {
          ...params,
          stream: true,
          contents,
        },
        options
      );

    let defaultRole: GeminiContentRole | undefined = 'model';
    for await (const content of stream) {
      const candidate: GenerateContentCandidate | undefined =
        content.candidates?.[0];

      if (!candidate) {
        continue;
      }

      const newMessage: BaseMessage = getMessageFromContentWithRole(
        candidate.content,
        defaultRole
      );

      defaultRole =
        (candidate.content.role as GeminiContentRole) ?? defaultRole;

      const generationChunk = new ChatGenerationChunk({
        output:
          typeof newMessage.content === 'string' ? newMessage.content : '',
        message: newMessage,
        info: {
          finishReason: candidate.finishReason,
          completion: candidate.index ?? 0,
          finishMessage: candidate.finishMessage,
          safetyRatings: candidate.safetyRatings,
          citationMetadata: candidate.citationMetadata,
        },
      });

      yield generationChunk;
    }

    if (options.signal?.aborted) {
      throw new Error('AbortError');
    }
  }

  private async *_generateContentIterable(
    model: GenerativeModel,
    request: GenerateContentRequest,
    options?: GeminiCallOptions
  ): AsyncIterable<EnhancedGenerateContentResponse> {
    const stream: AsyncGenerator<EnhancedGenerateContentResponse> =
      await this.caller.callWithOptions(
        { signal: options?.signal },
        async () => {
          const { stream, response } =
            await model.generateContentStream(request);

          const res = await response;
          if (
            res &&
            'promptFeedback' in res &&
            typeof res['promptFeedback'] === 'object' &&
            'blockReason' in res['promptFeedback']
          ) {
            this._throwErrorForBlockReason(
              res['promptFeedback']['blockReason']
            );
          }

          return stream;
        }
      );

    // Iterate over the async generator from the stream
    for await (const content of stream) {
      yield content;
    }
  }

  async completionWithRetry(
    request: GeminiChatParamsStreaming,
    options?: GeminiCallOptions
  ): Promise<AsyncIterable<GenerateContentResponse>>;

  async completionWithRetry(
    request: GeminiChatParamsNonStreaming,
    options?: GeminiCallOptions
  ): Promise<GenerateContentResponse>;

  async completionWithRetry(
    request: GeminiChatParamsNonStreaming | GeminiChatParamsStreaming,
    options?: GeminiCallOptions
  ): Promise<AsyncIterable<GenerateContentResponse> | GenerateContentResponse> {
    const modelParams: ModelParams = this._getModelParams(options);

    const model: GenerativeModel =
      this._getClient().getGenerativeModel(modelParams);

    const req: GenerateContentRequest = {
      contents: request.contents,
      safetySettings: request.safetySettings,
      generationConfig: request.generationConfig,
    };

    if (request.stream) {
      return this._generateContentIterable(model, req, options);
    } else {
      return await this.caller.call(async () => {
        const res: GenerateContentResult = await model.generateContent(req);

        const response = res.response;
        if (
          response &&
          'promptFeedback' in response &&
          typeof response['promptFeedback'] === 'object' &&
          response['promptFeedback'] !== null &&
          'blockReason' in response['promptFeedback']
        ) {
          this._throwErrorForBlockReason(
            response['promptFeedback']['blockReason'] as string
          );
        }

        return response;
      });
    }
  }

  protected _throwErrorForBlockReason(
    blockReason: string,
    blockReasonMessage?: string
  ) {
    const error = new Error(
      `The message is blocked because of ${blockReason}` +
        (blockReasonMessage ? `: ${blockReasonMessage}` : '')
    );

    (error as AsyncCallError).status = 406;

    throw error;
  }

  getModelContextSize(modelName: 'gemini-pro' | 'gemini-pro-vision'): number {
    switch (modelName) {
      case 'gemini-pro':
        return 30720;
      case 'gemini-pro-vision':
        return 12288;
    }
  }

  async getNumTokensInGenerations(generations: ChatGenerationChunk[]) {
    const geminiContents: Content[] = generations.map(
      (generation: ChatGenerationChunk) =>
        getContentFromMessage(generation.message)
    );

    return await this.getNumTokensInContents(geminiContents);
  }

  async getNumTokensInChat(messages: BaseMessage[]): Promise<number> {
    const geminiContents: Content[] = messages.map((message: BaseMessage) =>
      getContentFromMessage(message)
    );

    return await this.getNumTokensInContents(geminiContents);
  }

  async getNumTokensInContents(geminiContents: Content[]): Promise<number> {
    const numTokensResponse: CountTokensResponse = await this._getClient()
      .getGenerativeModel({
        model: this.modelName,
      })
      .countTokens({
        contents: geminiContents,
      });

    return numTokensResponse.totalTokens;
  }

  private _getClient(): GoogleGenerativeAIClient {
    if (!this._client) {
      if (!this.googleApiKey) {
        throw new Error('Google API Key not found');
      }

      this._client = new GoogleGenerativeAIClient(this.googleApiKey);
    }

    return this._client;
  }

  private _getModelParams(options?: GeminiCallOptions): ModelParams {
    return {
      model: this.modelName,
      // tools: this._getTools(options as this["SerializedCallOptions"]),
      safetySettings: this._getSafetySettings(
        options as this['SerializedCallOptions']
      ),
      generationConfig: this._getGenerationConfig(),
      ...this.additionalKwargs,
    };
  }
}
