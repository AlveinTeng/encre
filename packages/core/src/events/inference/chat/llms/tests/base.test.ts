import { expect, test } from '@jest/globals';
import { stringify } from 'yaml';
import { MemoryCache } from '../../../../../cache/index.js';
import { batch } from '../../../../../utils/batch.js';
import {
  BaseMessage,
  BaseMessageLike,
  BotMessage,
  HumanMessage,
  SystemMessage,
  convertMessageLikeToMessage,
  getChatString,
} from '../../../../input/load/msgs/base.js';
import {
  BasePrompt,
  StringPrompt,
} from '../../../../input/load/prompts/base.js';
import { ChatPrompt } from '../../../../input/load/prompts/chat.js';
import { Generation } from '../../../../output/provide/generation.js';
import { LLMResult } from '../../../../output/provide/llmresult.js';
import {
  BaseLLM,
  BaseLM,
  BaseLMCallOptions,
  BaseLMInput,
  calculateMaxToken,
} from '../base.js';

test('test calculateMaxToken', async () => {
  const prompt = 'this is a prompt.';

  const modelName = 'gpt-3.5-turbo-16k-0613';
  expect(await calculateMaxToken(prompt, modelName)).toBe(16384 - 5);

  const modelName2 = 'gpt-3.5-turbo-0613';
  expect(await calculateMaxToken(prompt, modelName2)).toBe(4096 - 5);

  const modelName3 = 'gpt-4-32k-0314';
  expect(await calculateMaxToken(prompt, modelName3)).toBe(32768 - 5);

  const modelName4 = 'gpt-4-0314';
  expect(await calculateMaxToken(prompt, modelName4)).toBe(8192 - 5);

  const modelName5 = 'gpt2';
  expect(await calculateMaxToken(prompt, modelName5)).toBe(4096 - 5);
});

test('test BaseLM', async () => {
  class TestLM extends BaseLM {
    _namespace: string[] = ['tests'];

    _llmType(): string {
      return 'test_llm';
    }

    _modelType(): string {
      return 'test_model';
    }

    /**
     * In this method, we mock the language model result as pairs of prompts
     * given an array of prompts.
     * @param prompts The given array of prompts in strings.
     * @param options if given an array of strings, then it means the stopwords.
     * @param callbacks TODO: not yet implemented
     */
    async provide(
      prompts: string[],
      options?: BaseLMCallOptions | string[],
      callbacks?: any
    ): Promise<LLMResult> {
      const generations: Generation[][] = batch(prompts, 2).map(
        (promptPair: string[]) =>
          promptPair.map((prompt: string, idx: number) => ({
            output: prompt,
            info: {
              index: idx,
            },
          }))
      );

      return { generations };
    }

    /**
     * In this method, we mock the language model result as pairs of prompts
     * given an array of prompts.
     * @param prompts The given array of prompts in {@link BasePrompt}.
     * @param options if given an array of strings, then it means the stopwords.
     * @param callbacks TODO: not yet implemented
     */
    async provideWithPrompts(
      prompts: BasePrompt[],
      options?: BaseLMCallOptions | string[],
      callbacks?: any
    ): Promise<LLMResult> {
      const promptStrs: string[] = prompts.map((prompt: BasePrompt) =>
        prompt.toString()
      );

      const generations: Generation[][] = batch(promptStrs, 2).map(
        (promptPair: string[]) =>
          promptPair.map((prompt: string, idx: number) => ({
            output: prompt,
            info: {
              index: idx,
            },
          }))
      );

      return { generations };
    }

    /**
     * In this method, we mock the predict text as the given text input.
     * @param text The given input in string.
     * @param options if given an array of strings, then it means the stopwords.
     * @param callbacks TODO: not yet implemented
     * @returns
     */
    async predict(
      text: string,
      options?: BaseLMCallOptions | string[],
      callbacks?: any
    ): Promise<string> {
      return text;
    }

    /**
     * In this method, we mock the predict text as the given text input in {@link BotMessage}.
     * @param messages The given input in {@link BaseMessage[]}.
     * @param options if given an array of strings, then it means the stopwords.
     * @param callbacks TODO: not yet implemented
     */
    async predictWithMessages(
      messages: BaseMessage[],
      options?: BaseLMCallOptions | string[] | undefined,
      callbacks?: any
    ): Promise<BaseMessage> {
      const text: string = getChatString(messages);
      return new BotMessage(text);
    }

    /**
     * Invokes the language model with a given input and options.
     * @param input The given prompt in {@link BaseLMInput}
     * @param options It contains callbask and other options such as stopwords.
     */
    async invoke(
      input: BaseLMInput,
      options?: Partial<BaseLMCallOptions>
    ): Promise<unknown> {
      let prompt: BasePrompt;
      if (typeof input === 'string') {
        prompt = new StringPrompt(input);
      } else if (Array.isArray(input)) {
        prompt = new ChatPrompt(
          input.map((msg: BaseMessageLike) => convertMessageLikeToMessage(msg))
        );
      } else {
        prompt = input;
      }

      const result: LLMResult = await this.provideWithPrompts(
        [prompt],
        options,
        options?.callbacks
      );

      return result;
    }
  }

  const testLM = new TestLM({});

  // BaseLM is not serializable
  const serializedStr: string = JSON.stringify(testLM, null, 2);
  expect(stringify(JSON.parse(serializedStr))).toMatchSnapshot();

  const result: string = await testLM.predict('this is a prompt.');
  expect(result).toBe('this is a prompt.');

  const messages: BaseMessage[] = [
    new SystemMessage('this is a system message'),
    new HumanMessage('this is a human message'),
  ];
  const result2: BaseMessage = await testLM.predictWithMessages(messages);
  expect(JSON.stringify(result2, null, 2)).toMatchSnapshot();

  const prompts: string[] = [
    'System: this is a system message',
    'Human: this is a human message',
    'AI: thhis is a bot message',
  ];
  const result3: LLMResult = await testLM.provide(prompts);
  expect(JSON.stringify(result3, null, 2)).toMatchSnapshot();

  const basePrompts: BasePrompt[] = [
    new StringPrompt('this is a prompt.'),
    new ChatPrompt(messages),
    new StringPrompt('this is a prompt.'),
  ];
  const result4: LLMResult = await testLM.provideWithPrompts(basePrompts);
  expect(JSON.stringify(result4, null, 2)).toMatchSnapshot();
});

test('test BaseLLM', async () => {
  class TestLLM extends BaseLLM {
    /**
     * In this method, we mock the language model result as pairs of prompts
     * given an array of prompts.
     * @param prompts The given array of prompts in strings.
     * @param options if given an array of strings, then it means the stopwords.
     * @param callbacks TODO: not yet implemented
     */
    async _provide(
      prompts: string[],
      options: this['SerializedCallOptions']
    ): Promise<LLMResult> {
      const generations: Generation[][] = batch(prompts, 2).map(
        (promptPair: string[]) =>
          promptPair.map((prompt: string, idx: number) => ({
            output: prompt,
            info: {
              index: idx,
            },
          }))
      );

      return { generations };
    }

    _llmType(): string {
      return 'test_llm';
    }
  }

  const cache: MemoryCache = MemoryCache.global();
  const testLLMWithCustomCache = new TestLLM({ cache });

  // BaseLLM is not serializable
  const serializedStr: string = JSON.stringify(testLLMWithCustomCache, null, 2);
  expect(stringify(JSON.parse(serializedStr))).toMatchSnapshot();

  expect(await testLLMWithCustomCache.predict('this is a prompt.')).toBe(
    'this is a prompt.'
  );

  const messages: BaseMessage[] = [
    new SystemMessage('this is a system message'),
    new HumanMessage('this is a human message'),
  ];
  const result2: BaseMessage =
    await testLLMWithCustomCache.predictWithMessages(messages);
  expect(JSON.stringify(result2, null, 2)).toMatchSnapshot();

  const prompts: string[] = [
    'System: this is a system message',
    'Human: this is a human message',
    'AI: thhis is a bot message',
  ];
  const result3: LLMResult = await testLLMWithCustomCache.provide(prompts);
  expect(JSON.stringify(result3, null, 2)).toMatchSnapshot();

  const basePrompts: BasePrompt[] = [
    new StringPrompt('this is a prompt.'),
    new ChatPrompt(messages),
    new StringPrompt('this is a prompt.'),
  ];
  const result4: LLMResult =
    await testLLMWithCustomCache.provideWithPrompts(basePrompts);
  expect(JSON.stringify(result4, null, 2)).toMatchSnapshot();
});
