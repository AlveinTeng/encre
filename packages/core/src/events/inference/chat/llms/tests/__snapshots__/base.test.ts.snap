// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test BaseLLM 1`] = `
"_grp: 1
_type: not_implemented
_id:
  - inference
  - chat
  - llms
  - test_llm
  - TestLLM
"
`;

exports[`test BaseLLM 2`] = `
"{
  "_grp": 1,
  "_type": "constructor",
  "_id": [
    "input",
    "load",
    "msgs",
    "BotMessage"
  ],
  "_kwargs": {
    "content": "System: this is a system message\\nHuman: this is a human message"
  }
}"
`;

exports[`test BaseLLM 3`] = `
"{
  "generations": [
    [
      {
        "output": "System: this is a system message",
        "info": {
          "index": 0
        }
      },
      {
        "output": "Human: this is a human message",
        "info": {
          "index": 1
        }
      }
    ],
    [
      {
        "output": "AI: this is a bot message",
        "info": {
          "index": 0
        }
      }
    ],
    null
  ],
  "llmOutput": {}
}"
`;

exports[`test BaseLLM 4`] = `
"{
  "generations": [
    [
      {
        "output": "this is a prompt.",
        "info": {
          "index": 0
        }
      }
    ],
    [
      {
        "output": "System: this is a system message\\nHuman: this is a human message",
        "info": {
          "index": 0
        }
      }
    ],
    [
      {
        "output": "this is a prompt.",
        "info": {
          "index": 0
        }
      }
    ]
  ],
  "llmOutput": {}
}"
`;

exports[`test BaseLLM 5`] = `
"this is a prompt.
"
`;

exports[`test BaseLLM 6`] = `
"|-
System: this is a system message
Human: this is a human message
AI: this is a bot message
"
`;

exports[`test BaseLLM 7`] = `
"|-
Human: this is a human message 1
Human: this is a human message 2
Human: this is a human message 3
"
`;

exports[`test BaseLLM 8`] = `
"|-
System: this is a system message
Human: this is a human message
"
`;

exports[`test BaseLM 1`] = `
"_grp: 1
_type: not_implemented
_id:
  - tests
  - TestLM
"
`;

exports[`test BaseLM 2`] = `
"{
  "_grp": 1,
  "_type": "constructor",
  "_id": [
    "input",
    "load",
    "msgs",
    "BotMessage"
  ],
  "_kwargs": {
    "content": "System: this is a system message\\nHuman: this is a human message"
  }
}"
`;

exports[`test BaseLM 3`] = `
"{
  "generations": [
    [
      {
        "output": "System: this is a system message",
        "info": {
          "index": 0
        }
      },
      {
        "output": "Human: this is a human message",
        "info": {
          "index": 1
        }
      }
    ],
    [
      {
        "output": "AI: this is a bot message",
        "info": {
          "index": 0
        }
      }
    ]
  ]
}"
`;

exports[`test BaseLM 4`] = `
"{
  "generations": [
    [
      {
        "output": "this is a prompt.",
        "info": {
          "index": 0
        }
      },
      {
        "output": "System: this is a system message\\nHuman: this is a human message",
        "info": {
          "index": 1
        }
      }
    ],
    [
      {
        "output": "this is a prompt.",
        "info": {
          "index": 0
        }
      }
    ]
  ]
}"
`;

exports[`test BaseLM 5`] = `
"generations:
  - - output: this is a prompt.
      info:
        index: 0
"
`;

exports[`test BaseLM 6`] = `
"generations:
  - - output: |-
        System: this is a system message
        Human: this is a human message
        AI: this is a bot message
      info:
        index: 0
"
`;

exports[`test BaseLM 7`] = `
"generations:
  - - output: |-
        Human: this is a human message 1
        Human: this is a human message 2
        Human: this is a human message 3
      info:
        index: 0
"
`;

exports[`test BaseLM 8`] = `
"generations:
  - - output: |-
        System: this is a system message
        Human: this is a human message
      info:
        index: 0
"
`;
